lib/compat.php

```diff
@@ -304,6 +304,247 @@ function gutenberg_warn_classic_about_blocks() {
+ * @return string $content Unchanged post content.
+ */
+function _gutenberg_cache_images_meta( $content ) {
+	// Need to find all image blocks and get the attachment IDs from them BEFORE the parser is run
+	// so we can get the image attachments meta all at once from the DB.
+	if ( preg_match_all( '/^<!-- wp:image {.*"id":(\d+),.*} -->$/m', $content, $matches ) ) {
+		_prime_post_caches( $matches[1], false, true );
+	}
+
+	return $content;
+}
+
+// Run before blocks are parsed.
+add_filter( 'the_content', '_gutenberg_cache_images_meta', 3 );
+
+/**
+ * Calculates the image width and height based on $block_witdh and the `editWidth` block attribute.
+ *
+ * @since 4.x.x
+ *
+ * @param array $block_attributes The block attributes.
+ * @param array $image_meta Optional. The image attachment meta data.
+ * @return array|bool An array of the image width and height, in that order, or false if the image data is missing from $block_attributes.
+ */
+function gutenberg_get_image_width_height( $block_attributes, $image_meta = null ) {
+	if ( ! empty( $block_attributes['width'] ) && ! empty( $block_attributes['height'] ) ) {
+		// The image was resized.
+		$image_dimensions = array(
+			$block_attributes['width'],
+			$block_attributes['height'],
+		);
+
+		/*
+		 * Here we can use `$block_attributes['editWidth']` to scale the image
+		 * if we know the theme's "expected width" (in pixels).
+		 *
+		 * Note that if the `$block_attributes['userSet']` is set/true, the user has entered
+		 * the width and height by hand, they shouldn't probably be changed.
+		 *
+		 * Something like:
+		 *	if ( empty( $block_attributes['userSet'] ) && ! empty( $block_attributes['editWidth'] ) && $content_width <> $block_attributes['editWidth'] ) {
+		 *		// Scale the image if the block width in the editor was different than the current theme width.
+		 *		$scale = $content_width / $block_attributes['editWidth'];
+		 *		$image_width = round( $block_attributes['width'] * $scale );
+		 *	}
+		 *	$image_dimensions = wp_constrain_dimensions( $image_file_width, $image_file_height, $image_width );
+		 */
+	} elseif ( ! empty( $block_attributes['fileWidth'] ) && ! empty( $block_attributes['fileHeight'] ) ) {
+		$image_dimensions = array(
+			$block_attributes['fileWidth'],
+			$block_attributes['fileHeight'],
+		);
+	} else {
+		return false;
+	}
+
+	/*
+	 * Do not constrain images with "wide" and "full" allignment to the "large" image size.
+	 * TODO: To reduce (fix) the need for upscaling or using the "full" size images
+	 * add "xlarge" image size generated by default!
+	 */
+	if ( ! empty( $image_meta ) &&
+		( 'wide' === $block_attributes['align'] || 'full' === $block_attributes['align'] ) &&
+		! empty( $block_attributes['fileWidth'] ) &&
+		$block_attributes['fileWidth'] < $image_meta['width'] &&
+		max( (int) $image_meta['width'], (int) $image_meta['height'] ) < 4300 && // max 12 MP photo (that's still may be over 3MB, consider reducing).
```

Comment:
This line gives me some pause because it introduces caveats to the claim that "Full wide alignment uses the original image", the number itself maybe (inherently) arbitrary, and very much a magic number (should be constant? filtered value?).

Reply:
Yeah, that's a pretty... arbitrary assumption. We just need a "sensible default" for now. 

Really hoping we can add another larger image size that's generated by default in core, like `xlarge`. Then we can just use it or fall back to "full" when it doesn't exist.

Reply:
Updated this to look for larger image sizes that were added by themes or plugins first, see https://github.com/WordPress/gutenberg/pull/11973/commits/44c7c68ae058792d65e88553f526df44358d5b32.